docs/source/tutorials/knn_outlier.ipynb

```diff
@@ -0,0 +1,447 @@
+    "trainX = trainX[animal_idxs]\n",
+    "trainy = trainy[animal_idxs]\n",
+    "\n",
+    "# Check the shapes of our training and test sets\n",
+    "print('Train: X=%s, y=%s' % (trainX.shape, trainy.shape))\n",
+    "print('Test: X=%s, y=%s' % (testX.shape, testy.shape))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d4480c85",
+   "metadata": {},
+   "source": [
+    "#### Lets visualize some of the training and test examples"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "94a7bb8a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "txt_labels = {0: 'airplane', \n",
+    "              1: 'automobile', \n",
+    "              2: 'bird',\n",
+    "              3: 'cat', \n",
+    "              4: 'deer', \n",
+    "              5: 'dog', \n",
+    "              6: 'frog', \n",
+    "              7: 'horse', \n",
+    "              8:'ship', \n",
+    "              9:'truck'}\n",
+    "\n",
+    "def plot_images(X,y):\n",
+    "    plt.rcParams[\"figure.figsize\"] = (9,7)\n",
+    "    for i in range(15):\n",
+    "        # define subplot\n",
+    "        ax = plt.subplot(3,5,i+1)\n",
+    "        ax.set_title(txt_labels[int(y[i])])\n",
+    "        # plot raw pixel data\n",
+    "        ax.imshow(X[i])\n",
+    "    # show the figure\n",
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9dddf313",
+   "metadata": {},
+   "source": [
+    "Observe how there are only animals left in the training set `trainX` below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "486acf25",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plot_images(trainX,trainy)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7cf11eb8",
+   "metadata": {},
+   "source": [
+    "The test set on the other hand still visibily contains the non-animal images: `[ship, airplane, automobile, truck]`. If we consider `trainX` to be the representation of data, then these non-animal images in test dataset `testX` become outliers."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1d2fdd09",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plot_images(testX,testy)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "15f57bae",
+   "metadata": {},
+   "source": [
+    "## 3. Import a model and embeddings\n",
+    "The model we are importing comes from [timm](https://timm.fast.ai/), a deep-learning library collection of SOTA models and utilities. We pass in the images into the model to generate embeddings in the feature space used in outlier detection.\n",
+    "\n",
+    "The model itself is a `resnet50` but outlier detection can be done with any object capable of generating feature embeddings."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "afd151d5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "import timm\n",
```

Comment:
move imports to start of tutorial, so if it fails, user can first install the requisite packages rather than only realizing half-way through that a dependency is missing.